File: routes/publisher/v2.js
"use strict";

const config = require('config');
const SERVER_CFG = config.get("SERVER");
const KEY = SERVER_CFG.api.KEY;

const express = require('express');
const router = express.Router();
const {
    validationResult
} = require('express-validator');

let logger = require("../../logger");

const ERRORS = require("../../error.js");
const Security = require("../../models/security.models.js");

const STORE_SESSION = require("../../models/store_sessions.models");
const FIND_USERS = require("../../models/find_users.models");
const VERIFY_PHONE_NUMBER = require("../../models/verify_phone_number.models");
const FIND_DEV_SESSION = require("../../models/find_developers_sessions.models");
const FIND_GRANTS = require("../../models/find_grant.models");
const DECRYPT_GRANTS = require("../../models/decrypt_grant.models");
const FIND_PLATFORMS = require("../../models/find_platforms.models");
const REFRESH_TOKENS = require("../../models/refresh_token.models");
const DELETE_GRANTS = require("../../models/delete_grant.models");
const PURGE_GRANTS = require("../../models/purge_grants.models");
const VERIFY_PLATFORMS = require("../../models/verify_platforms.models");

const VALIDATOR = require("../../models/validator.models");

// ==================== PUBLISHER ====================
router.post("/decrypt",
    VALIDATOR.phoneNumber,
    VALIDATOR.platform,
    VALIDATOR.developerCookies,
    VALIDATOR.userAgent,
    async (req, res) => {
        try {
            // Finds the validation errors in this request and wraps them in an object with handy functions
            const errors = validationResult(req);
            if (!errors.isEmpty()) {
                errors.array().map(err => {
                    if (err.param == "SWOBDev") {
                        logger.error(`${err.param}: ${err.msg}`);
                        throw new ERRORS.Unauthorized();
                    }
                    logger.error(`${err.param}: ${err.msg}`);
                });
                throw new ERRORS.BadRequest();
            }
            // =============================================================

            const PHONE_NUMBER = req.body.phone_number;
            const PLATFORM = req.body.platform;
            const USER_AGENT = req.get("user-agent");
            let DEV_COOKIE_CHECK = req.cookies.SWOBDev;
            let dev_cookie_buff = Buffer.from(DEV_COOKIE_CHECK, "base64")
            let dev_cookie_str = dev_cookie_buff.toString('utf-8');
            let dev_cookie = JSON.parse(dev_cookie_str)
            let DEV_COOKIE = dev_cookie.cookie;
            let DEV_USER_AGENT = dev_cookie.userAgent;
            let DEV_UID = dev_cookie.uid;
            let veri_path = dev_cookie.verification_path

            await FIND_DEV_SESSION(DEV_UID, DEV_USER_AGENT, DEV_COOKIE, veri_path);

            const USERID = await VERIFY_PHONE_NUMBER(PHONE_NUMBER);
            const USER = await FIND_USERS(USERID);
            let platform = await FIND_PLATFORMS(PLATFORM);
            const GRANT = await FIND_GRANTS(USER, platform);
            const DECRYPTED_GRANT = await DECRYPT_GRANTS(GRANT, USER);
            const refreshed_token = await REFRESH_TOKENS(DECRYPTED_GRANT.token, platform)

            if (refreshed_token) {
                let security = new Security(KEY);

                await GRANT.update({
                    username: security.encrypt(JSON.stringify(DECRYPTED_GRANT.username)).e_info,
                    uniqueId: security.encrypt(JSON.stringify(DECRYPTED_GRANT.uniqueId)).e_info,
                    token: security.encrypt(JSON.stringify(refreshed_token)).e_info,
                    iv: security.iv
                })
            }

            await STORE_SESSION(USERID, USER_AGENT, null, null, "publisher");

            res.status(200).json({
                username: DECRYPTED_GRANT.username,
                token: refreshed_token ? refreshed_token : DECRYPTED_GRANT.token,
                uniqueId: DECRYPTED_GRANT.uniqueId
            });
        } catch (err) {
            if (err instanceof ERRORS.BadRequest) {
                return res.status(400).send(err.message);
            } // 400
            if (err instanceof ERRORS.Forbidden) {
                return res.status(403).send(err.message);
            } // 403
            if (err instanceof ERRORS.Unauthorized) {
                return res.status(401).send(err.message);
            } // 401
            if (err instanceof ERRORS.Conflict) {
                return res.status(409).send(err.message);
            } // 409
            if (err instanceof ERRORS.NotFound) {
                return res.status(200).send([]);
            } // 404

            logger.error(err.stack || err);
            return res.status(500).send("internal server error");
        }
    });

router.post("/whoami",
    VALIDATOR.phoneNumber,
    VALIDATOR.developerCookies,
    VALIDATOR.userAgent,
    async (req, res) => {
        try {
            // Finds the validation errors in this request and wraps them in an object with handy functions
            const errors = validationResult(req);
            if (!errors.isEmpty()) {
                errors.array().map(err => {
                    if (err.param == "SWOBDev") {
                        logger.error(`${err.param}: ${err.msg}`);
                        throw new ERRORS.Unauthorized();
                    }
                    logger.error(`${err.param}: ${err.msg}`);
                });
                throw new ERRORS.BadRequest();
            }
            // =============================================================

            const PHONE_NUMBER = req.body.phone_number;
            const USER_AGENT = req.get("user-agent");
            let DEV_COOKIE_CHECK = req.cookies.SWOBDev;
            let dev_cookie_buff = Buffer.from(DEV_COOKIE_CHECK, "base64")
            let dev_cookie_str = dev_cookie_buff.toString('utf-8');
            let dev_cookie = JSON.parse(dev_cookie_str)
            let DEV_COOKIE = dev_cookie.cookie;
            let DEV_USER_AGENT = dev_cookie.userAgent;
            let DEV_UID = dev_cookie.uid;
            let veri_path = dev_cookie.verification_path

            await FIND_DEV_SESSION(DEV_UID, DEV_USER_AGENT, DEV_COOKIE, veri_path);

            const USERID = await VERIFY_PHONE_NUMBER(PHONE_NUMBER);

            let session = await STORE_SESSION(USERID, USER_AGENT, null, null, "publisher");

            res.status(200).json({
                user_id: session.uid
            });
        } catch (err) {
            if (err instanceof ERRORS.BadRequest) {
                return res.status(400).send(err.message);
            } // 400
            if (err instanceof ERRORS.Forbidden) {
                return res.status(403).send(err.message);
            } // 403
            if (err instanceof ERRORS.Unauthorized) {
                return res.status(401).send(err.message);
            } // 401
            if (err instanceof ERRORS.Conflict) {
                return res.status(409).send(err.message);
            } // 409
            if (err instanceof ERRORS.NotFound) {
                return res.status(200).send([]);
            } // 404

            logger.error(err.stack || err);
            return res.status(500).send("internal server error");
        }
    });

router.delete("/revoke_all",
    async (req, res) => {
        try {
            const originalURL = req.header("Origin");

            var User = db.users;

            let user = await User.findAll().catch(error => {
                logger.error("ERROR FINDING USER");
                throw new ERRORS.InternalServerError(error);
            })

            for (let i = 0; i < user.length; i++) {
                logger.info(`Revoking ${i+1} of ${user.length} accounts ...`)
                let GRANTS = await user[i].getWallets();

                for (let j = 0; j < GRANTS.length; j++) {
                    let PLATFORM = await VERIFY_PLATFORMS(GRANTS[j].platformId)
                    let GRANT = await PURGE_GRANTS(originalURL, PLATFORM.name, GRANTS[j], user[i]);
                    await DELETE_GRANTS(GRANT);
                };
            }

            return res.status(200).json();
        } catch (err) {
            if (err instanceof ERRORS.BadRequest) {
                return res.status(400).send(err.message);
            } // 400
            if (err instanceof ERRORS.Forbidden) {
                return res.status(403).send(err.message);
            } // 403
            if (err instanceof ERRORS.Unauthorized) {
                return res.status(401).send(err.message);
            } // 401
            if (err instanceof ERRORS.Conflict) {
                return res.status(409).send(err.message);
            } // 409
            if (err instanceof ERRORS.NotFound) {
                return res.status(404).send(err.message);
            } // 404

            logger.error(err.stack || err);
            return res.status(500).send("internal server error");
        }
    });

// =============================================================

module.exports = router;

File: routes/publisher/v2.py
import logging
logger = logging.getLogger(__name__)

# configurations
from Configs import baseConfig
config = baseConfig()
api = config["API"]
cookie_name = api['COOKIE_NAME']
dev_cookie_name = api['DEV_COOKIE_NAME']
enable_otp_counter = eval(config["OTP"]["ENABLE"])

from flask import Blueprint
from flask import request
from flask import Response
from flask import jsonify

from platforms import platform_switch

from security.cookie import Cookie
from security.data import Data

import json
import base64

from datetime import datetime
from datetime import timedelta

from models.grants import Grant_Model
from models.platforms import Platform_Model
from models.users import User_Model
from models.sessions import Session_Model
from models._2FA import OTP_Model

v2 = Blueprint("v2", __name__)

from werkzeug.exceptions import BadRequest
from werkzeug.exceptions import Conflict
from werkzeug.exceptions import Unauthorized
from werkzeug.exceptions import Forbidden
from werkzeug.exceptions import InternalServerError
from werkzeug.exceptions import TooManyRequests

@v2.route("/decrypt", methods=["POST"])
def decrypt():
    """
    """
    try:
        User = User_Model()
        Session = Session_Model()
        Grant = Grant_Model()
        Platform = Platform_Model()

        if not request.cookies.get(dev_cookie_name):
            logger.error("no dev cookie")
            raise Unauthorized()
        elif not request.headers.get("User-Agent"):
            logger.error("no user agent")
            raise BadRequest()
        elif not "phone_number" in request.json or not request.json["phone_number"]:
            logger.error("no phone_number")
            raise BadRequest()
        elif not "platform" in request.json or not request.json["platform"]:
            logger.error("no platform")
            raise BadRequest()

        dev_cookie = json.loads(base64.b64decode(request.cookies.get(dev_cookie_name)))
        user_agent = request.headers.get("User-Agent")
        dev_uid = dev_cookie["uid"]
        dev_user_agent = dev_cookie["userAgent"]
        dev_verification_path = dev_cookie["verification_path"]

        phone_number = request.json["phone_number"]
        platform_name = request.json["platform"]

        Session.authenticate(
            uid=dev_uid,
            user_agent=dev_user_agent,
            cookie=dev_cookie["cookie"],
            verification_path=dev_verification_path
        )
    
        user = User.find(phone_number=phone_number)
        platform = Platform.find(platform_name=platform_name)
        
        try:
            grant = Grant.find(user_id=user["userId"], platform_id=platform.id)
            d_grant = Grant.decrypt(platform_name=platform.name, grant=grant, refresh=True)
        except BadRequest:
            d_grant = []

        res = jsonify(d_grant)

        Session.create(
            unique_identifier=user["userId"],
            user_agent=user_agent,
            type="publisher"
        )

        return res, 200
                
    except BadRequest as err:
        return str(err), 400

    except Unauthorized as err:
        return str(err), 401

    except Conflict as err:
        return str(err), 409

    except InternalServerError as err:
        logger.exception(err)
        return "internal server error", 500

    except Exception as err:
        logger.exception(err)
        return "internal server error", 500

@v2.route("/whoami", methods=["POST"])
def whoami():
    """
    """
    try:
        User = User_Model()
        Session = Session_Model()

        if not request.cookies.get(dev_cookie_name):
            logger.error("no dev cookie")
            raise Unauthorized()
        elif not request.headers.get("User-Agent"):
            logger.error("no user agent")
            raise BadRequest()
        elif not "phone_number" in request.json or not request.json["phone_number"]:
            logger.error("no phone_number")
            raise BadRequest()

        dev_cookie = json.loads(base64.b64decode(request.cookies.get(dev_cookie_name)))
        user_agent = request.headers.get("User-Agent")
        dev_uid = dev_cookie["uid"]
        dev_user_agent = dev_cookie["userAgent"]
        dev_verification_path = dev_cookie["verification_path"]

        phone_number = request.json["phone_number"]

        Session.authenticate(
            uid=dev_uid,
            user_agent=dev_user_agent,
            cookie=dev_cookie["cookie"],
            verification_path=dev_verification_path
        )
    
        user = User.find(phone_number=phone_number)

        res = jsonify({
            "user_id": user["userId"]
        })

        return res, 200
                
    except BadRequest as err:
        return str(err), 400

    except Unauthorized as err:
        return str(err), 401

    except Conflict as err:
        return str(err), 409

    except InternalServerError as err:
        logger.exception(err)
        return "internal server error", 500

    except Exception as err:
        logger.exception(err)
        return "internal server error", 500

